### PR TITLE
Allow more options than tax zone for price calculation

### PR DESCRIPTION
**Pull request by mamhoff**
Our client needs to display different prices to different customers. The MOSS
refactoring in https://github.com/spree/spree/commit/a1172606f27ee2e71f097bf301df9f99881ad2f5 has enabled different prices for different tax zones, which is great,
but they also need to display prices excluding VAT for _some_ business customers.

The logic for that is (while in accordance with EU tax law) rather custom. This PR
is supposed only to enable passing more than just the tax zone as options for price calculation.

In order to accomplish this, the `current_tax_zone` helper is replaced with a `current_price_options` helper which returns a Hash. All the values of that Hash are used to calculate product cache keys. I took care to order those so that we don't end up with different caches for the same content.

The changes in this PR comprise the following:
- Remove current_tax_zone helper and replace with current_price_options
- Adapt product cache key to price options hash
- Adapt base helper to call #price_including_vat_for with price_options hash
- Adapt price model for price_options hash
- Adapt VAT price calculation module
- Use new #included_tax_amount signature
- Adapt tax_rate.rb
- Adapt line_item.rb
- Adapt stock_estimator.rb
- Add tests for edge case price options

FYI: With this branch, you can override the `current_price_options` helper and the `VatPriceCalculation#gross_amount` to customize how prices are displayed. You'll also have to subclass the `DefaultTaxCalculator` to customize how taxes are calculated on your shiny custom prices, and you'll have to edit `LineItem#update_price` to also be in accordance. We don't see all these changes in the Spree core, but without this PR, that kind of logic would be very difficult to implement.

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://magiclabs-feature-allow-more-price-options.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://magiclabs-feature-allow-more-price-options.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: core/lib/spree/core/controller_helpers/store.rb [hunk 1]
```diff
@@ -18,6 +18,14 @@ def current_store
           @current_store ||= Spree::Store.current(request.env['SERVER_NAME'])
         end
 
+        def current_price_options
```
Comment (JDutil) [inline core/lib/spree/core/controller_helpers/store.rb]: It would be good to add some documentation about this method here.

**Comment by JDutil** [comment]
I think it would be good to add this to the release notes as well, and add some documentation explaining the reason/usage of `current_price_options `.

**Comment by mamhoff** [comment]
Will do. I just checked, the previous `current_tax_zone` helper hasn't made it into a release yet, so I'll document this as a new helper.

**Comment by mamhoff** [comment]
Please don't merge as of yet, I found issues with Shipments that I didn't expect.

**Comment by mamhoff** [comment]
Solved the shipments issue. @JDutil This can be merged.

**Comment by mamhoff** [comment]
...rebased on master.

File: core/lib/spree/core/controller_helpers/store.rb [hunk 2]
```diff
@@ -18,6 +18,20 @@ def current_store
           @current_store ||= Spree::Store.current(request.env['SERVER_NAME'])
         end
 
+        # Return a Hash of things that influence the prices displayed in your shop.
+        #
+        # By default, the only thing that influences prices that is the current order's +tax_zone+
+        # (to facilitate differing prices depending on VAT rate for digital products in Europe, see
+        # https://github.com/spree/spree/pull/6295 and https://github.com/spree/spree/pull/6662).
+        #
```
Comment (JDutil) [inline core/lib/spree/core/controller_helpers/store.rb]: Could you please also include this last bit from the release notes in the comments here as well:
```
If your prices depend on something else, overwrite this method and add more key/value pairs to the Hash it returns.

Be careful though to also patch the following parts of Spree accordingly:

* `Spree::VatPriceCalculation#gross_amount`
* `Spree::LineItem#update_price`
* `Spree::Stock::Estimator#taxation_options_for`
* Subclass the `DefaultTax` calculator
```

**Comment by mamhoff** [comment]
Sure, here you go. There's actually a blog post coming up at how to configure Spree for complex taxation setups, I just don't know when I'll get to it.

**Comment by JDutil** [comment]
Whenever you finish the post perhaps update the comment and/or release notes with a link to it.  This looks good now. Waiting on specs.
